### ISSUE TITLE
Document EntityRepository to help contributors understand how to use it to add new entities

### ISSUE THREAD

**Issue by pmbrull** [open-metadata/OpenMetadata#1419]
While reviewing the backend implementation and its logic flow I had some trouble understanding the different steps that happened.

In the end, one of the issues I was getting was in misunderstanding the usage of `validate` and `storeRelationship` in the `Repository` classes.

I found it a bit troubling that in the `validate`, we are setting relationships and properties. As we are updating the input argument, it was a bit hard to follow 😅  (I am a bit more used to Scala's functional approach!)

For what I understood, in `dao.createOrUpdate` we are running `validate` (with under the hood setters) and then running `update` and `store`. However, `storeRelationship` is only called at `createNewEntity`.

Would it make sense to have a common approach wich a `validate -> set -> update -> store -> storeRel` to unify a bit the flow?

Also, do we have any docs explaining a bit the internals of the app? I think that would help in being more transparent in the architecture and design. I am lately receiving some feedback that the design is a bit hard to follow, and after some time working on this myself, I also have big knowledge gaps.

Sorry for the generic discussion here, let me know what you think @harshach @sureshms 🙏 

Many thanks!

**Comment by sureshms** [comment]
@pmbrull, these are fair questions that I myself have been thinking about. When we moved lot of common logic into a base class to delete repeated logic, the complexity of the code has increased. Few comments:
In `validate` - we validate both the attributes and the relationships of an entity. At the end of this, the entity is ready to go with all the fields initialized. This should be called for POST, PUT, and PATCH operation. When validating relationships, we have complete information about the relationships that we are using to set up the information and we don't want to do it again in another method. There are couple of options here - document this behavior, change the name to `validateAndSet`, change the name to `set`. I prefer `validate` with clear documentation. Wdyt?

As regards to create and update, they are fundamentally different operations. In create, you are creating everything. In the other, you have to compare what exists, remove relationships and add new relationships. That is why we are using Updaters.

Here is the documentation for the class and the methods. Let me know if it adds clarity.

```
/**
 * This class is used by Entity Resources to perform READ and WRITE operations to the backend database to Create,
 * Retrieve, Update, and Delete entities.
 *
 * An entity has two types of fields - `attributes` and `relationships`. The `attributes` are the core properties of
 * the entity, example - entity id, name, fullyQualifiedName, columns for a table, etc. The `relationships` are
 * an associated between two entities, example - table belongs to a database, table has a tag, user owns a table, etc.
 * All relationships are captured using {@code EntityReference}.
 * 
 * Entities are stored as JSON documents in the database. Each entity is stored in a separate table and is accessed
 * through a <i>Data Access Object</i> or <i>DAO</i> that corresponds to each of the entity. All DAO objects for an
 * entity are available in {@code daoCollection}.
 * 
 * Relationships between entity is stored in a separate table that captures the edge - fromEntity, toEntity, and 
 * the relationship name.
 *
 * JSON document of an entity stores only required attributes of an entity. Some attributes such as <i>href</i>
 * are not stored and are created on the fly.
 *
 * Json document of an entity does not store relationships. As an example, JSON document for <i>table</i> entity
 * does not store the relationship <i>database</i> which is of type <i>EntityReference</i>. This is always retrieved
 * from the the relationship edges when required to ensure, the data stored is consistent and information in
 * responses is not stale.
 */


....

  /**
   * This method is used for validating an entity to be created during POST, PUT, and PATCH operations and set up the
   * entity with all the required attributes and relationships.
   *
   * The implementation of this method must perform the following:
   * <ol>
   *   <li>Set the values for attributes that are not required in the request but can be derived on the server side.
   *   Example - <i>>FullyQualifiedNames</i> of an entity can be derived from the hierarchy that an entity belongs to
   *   .</li>
   *   <li>Validate all the attributes of an entity.</li>
   *   <li>Validate all the relationships of an entity. As an example - during <i>table</i> creation,
   *   relationships such as <i>Tags</i>, <i>Owner</i>, <i>Database</i>a table belongs to are validated.
   *   During validation additional information that is not required in the create/update request are set up
   *   in the corresponding relationship fields.</li>
   * </ol>
   *
   * At the end of this operation, entity is expected to be valid and setup with all the fields that will be sent as
   * payload in the POST, PUT, and PATCH operations response.
   *
   * @see TableRepository#validate(Table) for an example implementation
   */
  public abstract void validate(T entity) throws IOException;

  /**
   * An entity is stored in the backend database as JSON document. The JSON includes only some of the attributes of the
   * entity and does not include attributes such as <i>href</i>. The relationship fields of an entity is never stored
   * in the JSON document. It is always reconstructed based on relationship edges from the backend database.
   *
   * As an example, when <i>table</i> entity is stored, the attributes such as <i>href</i> and the relationships such
   * as <i>owner</i>, <i>database</i>, and <i>tags</i> are set to null. These attributes are restored back after the
   * JSON document is stored to be sent as response.
   *
   * @see TableRepository#store(Table, boolean) for an example implementation
   */
  public abstract void store(T entity, boolean update) throws IOException;

  /**
   * This method is called to store all the relationships of an entity. It is expected that all relationships are
   * already validated and completely setup before this method is called and no validation of relationships is required.
   *
   * @see TableRepository#storeRelationships(Table) for an example implementation
   */
  public abstract void storeRelationships(T entity) throws IOException;
```

**Comment by sureshms** [comment]
Another idea, we can also call `validate` as `prepare`.

**Comment by pmbrull** [comment]
Thanks, @sureshms for the explanation and for passing this doc. It is much clearer now. With proper docs, I believe we can stick with `validate`.

I'll try to spend more time reading the interfaces to grasp all the internal behavior.

Thanks 🙏

**Comment by pmbrull** [comment]
I'll close the issue as the information is currently already available. We can plan an epic for the future to prepare logic flows in diagrams and recap the different internals in the docs if needed. Thanks!

**Comment by sureshms** [comment]
Reopening the issue to add documentation posted in the comments
